src/satosa/backends/idpy_oidc.py

```diff
@@ -0,0 +1,144 @@
+"""
+OIDC/OAuth2 backend module.
+"""
+from datetime import datetime
+import logging
+from urllib.parse import urlparse
+
+from idpyoidc.client.oauth2.stand_alone_client import StandAloneClient
+from idpyoidc.server.user_authn.authn_context import UNSPECIFIED
+
+from satosa.backends.base import BackendModule
+from satosa.internal import AuthenticationInformation
+from satosa.internal import InternalData
+import satosa.logging_util as lu
+from ..exception import SATOSAAuthenticationError
+from ..exception import SATOSAError
+from ..response import Redirect
+
+logger = logging.getLogger(__name__)
+
+
+class IdpyOIDCBackend(BackendModule):
+    """
+    Backend module for OIDC and OAuth 2.0, can be directly used.
+    """
+
+    def __init__(self, auth_callback_func, internal_attributes, config, base_url, name):
+        """
+        OIDC backend module.
+        :param auth_callback_func: Callback should be called by the module after the authorization
+        in the backend is done.
+        :param internal_attributes: Mapping dictionary between SATOSA internal attribute names and
+        the names returned by underlying IdP's/OP's as well as what attributes the calling SP's and
+        RP's expects namevice.
+        :param config: Configuration parameters for the module.
+        :param base_url: base url of the service
+        :param name: name of the plugin
+
+        :type auth_callback_func:
+        (satosa.context.Context, satosa.internal.InternalData) -> satosa.response.Response
+        :type internal_attributes: dict[string, dict[str, str | list[str]]]
+        :type config: dict[str, dict[str, str] | list[str]]
+        :type base_url: str
+        :type name: str
+        """
+        super().__init__(auth_callback_func, internal_attributes, base_url, name)
+        # self.auth_callback_func = auth_callback_func
+        # self.config = config
+        self.client = StandAloneClient(config=config["client"], client_type="oidc")
+        self.client.do_provider_info()
+        self.client.do_client_registration()
+
+    def start_auth(self, context, internal_request):
+        """
+        See super class method satosa.backends.base#start_auth
+
+        :type context: satosa.context.Context
+        :type internal_request: satosa.internal.InternalData
+        :rtype satosa.response.Redirect
+        """
+        login_url = self.client.init_authorization()
```

Comment:
The login_url will be the same always. Then this should be calculated once during initialization (`__init__`).

Reply:
As the line looks like right now that is true.
I don't know if context and/or internal request which are both arguments to the call could have any influence on what the authorisation request should look like.

As it's written right now that only things that matter is client configuration and the provider information and if this is how it should be then you're right.